src/sync/sync_user.hpp

```diff
@@ -0,0 +1,117 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYNC_USER_HPP
+#define REALM_OS_SYNC_USER_HPP
+
+#include <string>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+#include <mutex>
+
+namespace realm {
+
+class SyncSession;
+
+// A `SyncUser` represents a single user account. Each user manages the sessions that
+// are associated with it.
+class SyncUser {
+friend class SyncSession;
+public:	
+    enum class State {
+        LoggedOut,
+        Active,
+        Error,
+    };
+
+    // Don't use this directly; use the `SyncManager` APIs. Public for use with `make_shared`.
+    SyncUser(std::string refresh_token, std::string identity, bool is_admin=false);
+
+    // Return a list of all sessions belonging to this user.
+    std::vector<std::shared_ptr<SyncSession>> all_sessions();
+
+    // Return a session for a given URL.
+    std::shared_ptr<SyncSession> session_for_url(const std::string& url);
+
+    // Update the user's refresh token. If the user is logged out, it will log itself back in.
+    void update_refresh_token(std::string token);
+
+    // Log the user out and mark it as invalid. This will also close its associated Sessions.
+    void log_out();
+
+    // Whether the user was configured as an 'admin user' (directly uses its user token
+    // to open Realms).
+    bool is_admin() const
+    {
+        return m_is_admin;
+    }
+
+    std::string identity() const
+    {
+        return m_identity;
```

Comment:
`is_admin` and `identity` are internally set at creation and considered immutable. `state` should be protected by the lock.